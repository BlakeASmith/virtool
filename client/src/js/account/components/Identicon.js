/**
 *
 *
 * @copyright 2017 Government of Canada
 * @license MIT
 * @author igboyes
 *
 */
import React from "react";

const hsl2rgb = (h, s, b) => {
    h *= 6;
    s = [
        b += s *= b < .5 ? b : 1 - b,
        b - h % 1 * s * 2,
        b -= s *= 2,
        b,
        b + h % 1 * s,
        b + s
    ];

    return[
        Math.floor(s[ ~~h    % 6 ] * 255), // red
        Math.floor(s[ (h|16) % 6 ] * 255), // green
        Math.floor(s[ (h|8)  % 6 ] * 255),  // blue
        255
    ];
};

const Identicon = ({ hash }) => {

    const size = 64;
    const cell = Math.floor(64 / 5);

    const hue = parseInt(hash.substr(-7), 16) / 0xfffffff;
    const saturation = 0.7;
    const brightness = 0.5;

    let foreground = hsl2rgb(saturation, brightness, hue);
    let background = [240, 240, 240, 255];

    foreground = `rgba(${foreground.join(",")})`;
    background = `rgba(${background.join(",")})`;

    let rectPropsArray = [];

    for (let i = 0; i < 15; i++) {
        const color = parseInt(hash.charAt(i), 16) % 2 ? background: foreground;

        if (i < 5) {
            rectPropsArray.push({key: i, x: 2 * cell, y: i * cell, fill: color});
        } else if (i < 10) {
            const y = (i - 5) * cell;
            rectPropsArray.push({key: `${i}_0`, x: cell, y, fill: color});
            rectPropsArray.push({key: `${i}_1`, x: 3 * cell, y, fill: color});
        } else if (i < 15) {
            const y = (i - 10) * cell;
            rectPropsArray.push({key: `${i}_0`, x: 0, y, fill: color});
            rectPropsArray.push({key: `${i}_1`, x: 4 * cell, y, fill: color});
        }
    }

    return (
        <svg width={size} height={size}>
            {rectPropsArray.map(rectProps => <rect height={cell} width={cell} {...rectProps} />)}
        </svg>
    );
};

export default Identicon;
