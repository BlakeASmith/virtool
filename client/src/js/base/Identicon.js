import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import Identiconjs from "identicon.js";

const StyledIdenticon = styled.img`
    border-radius: ${props => props.size / 2}px;
    box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06);
    height: ${props => props.size}px;
    width: ${props => props.size}px;
`;

/**
 * Generates an identicon SVG from a user's identicon hash. Used for visually identifying users.
 *
 * @param size {number} the size of SVG to render
 * @param hash {string} the users identicon hash
 */
export const Identicon = ({ size = 64, hash }) => {
    const data = new Identiconjs(hash, {
        size,
        format: "svg"
    });

    return <StyledIdenticon src={`data:image/svg+xml;base64,${data}`} size={size} />;
};

Identicon.propTypes = {
    size: PropTypes.number,
    hash: PropTypes.string.isRequired
};
