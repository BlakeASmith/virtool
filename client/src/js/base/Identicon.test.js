import { Identicon } from "./Identicon";

describe("<Identicon />", () => {
    const size = 64;
    const hash = "d3b07384d113edec49eaa6238ad5ff00";

    const wrapper = shallow(<Identicon size={size} hash={hash} />);

    it("renders correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("renders an <img> element", () => {
        expect(wrapper.type()).toBe("img");
    });

});
