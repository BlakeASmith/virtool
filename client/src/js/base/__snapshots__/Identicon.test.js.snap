// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Identicon /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Identicon
    hash="d3b07384d113edec49eaa6238ad5ff00"
    size={64}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "height": 64,
      "src": "data:image/svg+xml;base64,PHN2ZyB4bWxucz0naHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmcnIHdpZHRoPSc2NCcgaGVpZ2h0PSc2NCcgc3R5bGU9J2JhY2tncm91bmQtY29sb3I6cmdiYSgyNDAsMjQwLDI0MCwxKTsnPjxnIHN0eWxlPSdmaWxsOnJnYmEoNTAsMzgsMjE3LDEpOyBzdHJva2U6cmdiYSg1MCwzOCwyMTcsMSk7IHN0cm9rZS13aWR0aDowLjMyOyc+PHJlY3QgIHg9JzI3JyB5PSczNycgd2lkdGg9JzEwJyBoZWlnaHQ9JzEwJy8+PHJlY3QgIHg9JzE3JyB5PScxNycgd2lkdGg9JzEwJyBoZWlnaHQ9JzEwJy8+PHJlY3QgIHg9JzM3JyB5PScxNycgd2lkdGg9JzEwJyBoZWlnaHQ9JzEwJy8+PHJlY3QgIHg9JzE3JyB5PScyNycgd2lkdGg9JzEwJyBoZWlnaHQ9JzEwJy8+PHJlY3QgIHg9JzM3JyB5PScyNycgd2lkdGg9JzEwJyBoZWlnaHQ9JzEwJy8+PHJlY3QgIHg9JzcnIHk9JzI3JyB3aWR0aD0nMTAnIGhlaWdodD0nMTAnLz48cmVjdCAgeD0nNDcnIHk9JzI3JyB3aWR0aD0nMTAnIGhlaWdodD0nMTAnLz48cmVjdCAgeD0nNycgeT0nNDcnIHdpZHRoPScxMCcgaGVpZ2h0PScxMCcvPjxyZWN0ICB4PSc0NycgeT0nNDcnIHdpZHRoPScxMCcgaGVpZ2h0PScxMCcvPjwvZz48L3N2Zz4=",
      "width": 64,
    },
    "ref": null,
    "rendered": null,
    "type": "img",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "height": 64,
        "src": "data:image/svg+xml;base64,PHN2ZyB4bWxucz0naHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmcnIHdpZHRoPSc2NCcgaGVpZ2h0PSc2NCcgc3R5bGU9J2JhY2tncm91bmQtY29sb3I6cmdiYSgyNDAsMjQwLDI0MCwxKTsnPjxnIHN0eWxlPSdmaWxsOnJnYmEoNTAsMzgsMjE3LDEpOyBzdHJva2U6cmdiYSg1MCwzOCwyMTcsMSk7IHN0cm9rZS13aWR0aDowLjMyOyc+PHJlY3QgIHg9JzI3JyB5PSczNycgd2lkdGg9JzEwJyBoZWlnaHQ9JzEwJy8+PHJlY3QgIHg9JzE3JyB5PScxNycgd2lkdGg9JzEwJyBoZWlnaHQ9JzEwJy8+PHJlY3QgIHg9JzM3JyB5PScxNycgd2lkdGg9JzEwJyBoZWlnaHQ9JzEwJy8+PHJlY3QgIHg9JzE3JyB5PScyNycgd2lkdGg9JzEwJyBoZWlnaHQ9JzEwJy8+PHJlY3QgIHg9JzM3JyB5PScyNycgd2lkdGg9JzEwJyBoZWlnaHQ9JzEwJy8+PHJlY3QgIHg9JzcnIHk9JzI3JyB3aWR0aD0nMTAnIGhlaWdodD0nMTAnLz48cmVjdCAgeD0nNDcnIHk9JzI3JyB3aWR0aD0nMTAnIGhlaWdodD0nMTAnLz48cmVjdCAgeD0nNycgeT0nNDcnIHdpZHRoPScxMCcgaGVpZ2h0PScxMCcvPjxyZWN0ICB4PSc0NycgeT0nNDcnIHdpZHRoPScxMCcgaGVpZ2h0PScxMCcvPjwvZz48L3N2Zz4=",
        "width": 64,
      },
      "ref": null,
      "rendered": null,
      "type": "img",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
